src/DFA/DFA.fs

```diff
@@ -106,84 +66,69 @@ type DFA<'a>(startState: State<'a>) =
-
-            let copyOfState = Dictionary<_, _>()
-
-            while toProcess.Count > 0 do
-                let state1, state2, addTransitions = toProcess.Pop()
-                let exists, newState = copyOfState.TryGetValue((state1, state2))
-
-                if not exists then
-                    let newState =
-                        State<_>(
-                            $"{state1.Id}_{state2.Id}",
-                            state1.IsStart && state2.IsStart,
-                            state1.IsFinal && state2.IsFinal
-                        )
-
-                    if newState.IsStart then
-                        newStartState <- newState
-
-                    addTransitions newState
-
-                    for (char1, targetState1) in state1.GetAllTransitions() do
-                        for (char2, targetState2) in state2.GetAllTransitions() do
-                            if char1 = char2 then
-                                toProcess.Push(
-                                    targetState1,
-                                    targetState2,
-                                    (fun newTargetState -> newState.AddTransition char1 newTargetState)
-                                )
-
-                    copyOfState.Add((state1, state2), newState)
-                else
-                    addTransitions newState
-
-            newStartState
-
-        let newStartState = dfs dfa1.StartState dfa2.StartState
-        DFA(newStartState)
-
-    let union dfa1 dfa2 =
+    let complement (dfa: DFA) =
+        DFA(
+            dfa.StartState,
+            dfa.Transitions
+            |> Seq.map (fun node -> node.Key)
+            |> Seq.filter (not << (fun x -> Seq.contains x dfa.FinalStates))
+            |> Set.ofSeq,
+            dfa.Transitions
+        )
+
+    let intersection (dfa1: DFA) (dfa2: DFA) =
+        let stateNumber2 = dfa2.Transitions.Count
+
+        let makeNewState (state1, state2) =
+            { Name = $"{state1.Name}_{state2.Name}"
+              Id = state1.Id * stateNumber2 + state2.Id }
+
+        let makeNewStates (fst, snd) =
+            Seq.allPairs fst snd |> Seq.map makeNewState
+
+        let startState = makeNewState (dfa1.StartState, dfa2.StartState)
+
+        let transitions =
+            Seq.allPairs dfa1.Transitions dfa2.Transitions
+            |> Seq.map (fun (first, second) ->
+                (makeNewState (first.Key, second.Key),
+                 first.Value
+                 |> Seq.map (fun node -> (node.Key, makeNewStates (node.Value, second.Value[node.Key])))
+                 |> Map.ofSeq))
+            |> Map.ofSeq
+
+        let finalStates =
+            Seq.allPairs dfa1.FinalStates dfa2.FinalStates
+            |> Seq.map makeNewState
+            |> Set.ofSeq
+
+        DFA(startState, finalStates, transitions)
+
+    let union (dfa1: DFA) (dfa2: DFA) =
         complement (intersection (complement dfa1) (complement dfa2))
```

Comment:
NFAs are not closed under complement, so we can't write union like that. But it'll probably be intersection with minor changes in requirements for final states